File: packages/client-core/src/state/reducer.ts
/** Ordered single-writer quote-state reducer. */
import {
  BPS,
  LaneFlags,
  quote as computeQuote,
  type Address,
  type LaneState,
  type QuoteOutcome,
  type QuoteRequest,
  type QuoteState,
} from "@lunarbase/math";
import {
  Commitment,
  commitmentRank,
  IndexerError,
  MATH_COMPATIBILITY_VERSION,
  ReducerError,
  SCHEMA_VERSION,
} from "../model.js";
import type { ChainCursor, Checkpoint, DeploymentConfig, QuoteEvent } from "../model.js";
import { compareCursor } from "../source.js";

const U128_MAX = (1n << 128n) - 1n;
const key = (address: Address): Address => address.toLowerCase() as Address;

function emptyLane(): LaneState {
  return {
    slot0: 0n,
    totalPrincipalAmount: 0n,
    slippageKBps: 0,
    blockDelay: 0,
    flags: 0,
  };
}

function cloneState(state: QuoteState): QuoteState {
  return {
    cash: state.cash,
    lanes: new Map([...state.lanes].map(([asset, lane]) => [key(asset), { ...lane }])),
    feeProfile: {
      whitelisted: state.feeProfile.whitelisted,
      blacklistFeeMultiplier: state.feeProfile.blacklistFeeMultiplier,
      partnerFeeBps: new Map([...state.feeProfile.partnerFeeBps].map(([asset, fee]) => [key(asset), fee])),
    },
  };
}

/** In-memory reducer whose maps never escape the client API. */
export class QuoteReducer {
  /** Complete quote-critical state mutated only by this ordered reducer. */
  private state: QuoteState;
  /** Last normalized head or event position accepted by the reducer. */
  private cursorValue?: ChainCursor;
  /** Fail-closed publication flag cleared by gaps and reducer failures. */
  private ready = false;

  /** Creates a not-ready reducer for one configured router. */
  constructor(
    state: QuoteState,
    /** Only router whose fee and whitelist updates affect this instance. */
    private readonly configuredRouter: Address,
  ) {
    this.state = cloneState(state);
  }

  /** Restores a prevalidated checkpoint. */
  static fromCheckpoint(checkpoint: Checkpoint): QuoteReducer {
    const reducer = new QuoteReducer(checkpoint.state, checkpoint.router);
    reducer.cursorValue = { ...checkpoint.cursor };
    reducer.ready = true;
    return reducer;
  }

  /** Returns the last accepted cursor. */
  cursor(): ChainCursor | undefined {
    return this.cursorValue ? { ...this.cursorValue } : undefined;
  }

  /** Reports whether quotes may use current state. */
  isReady(): boolean {
    return this.ready;
  }

  /** Revokes readiness until a complete canonical recovery succeeds. */
  markNotReady(): void {
    this.ready = false;
  }

  /** Publishes recovered state. */
  publishReady(): void {
    this.ready = true;
  }

  /** Installs the initial block-tagged cursor. */
  bootstrap(cursor: ChainCursor): void {
    this.cursorValue = { ...cursor };
    this.ready = true;
  }

  /** Advances head metadata without changing quote state. */
  observeHead(head: ChainCursor): void {
    const current = this.cursorValue;
    if (!current) {
      this.cursorValue = { ...head };
      return;
    }
    this.validateChainAndHash(current, head);
    if (head.blockNumber < current.blockNumber) return;
    if (head.blockNumber > current.blockNumber) {
      this.cursorValue = { ...head };
      return;
    }
    const progression = isRealtimeProgression(current, head);
    this.cursorValue = {
      ...current,
      executionBlockNumber: head.executionBlockNumber,
      blockHash: current.blockHash === undefined || progression ? head.blockHash : current.blockHash,
      commitment:
        commitmentRank(head.commitment) > commitmentRank(current.commitment) ? head.commitment : current.commitment,
      sourceSequence:
        (head.sourceSequence ?? 0n) > (current.sourceSequence ?? 0n) ? head.sourceSequence : current.sourceSequence,
      sourceSubIndex:
        (head.sourceSequence ?? 0n) > (current.sourceSequence ?? 0n) ? head.sourceSubIndex : current.sourceSubIndex,
    };
  }

  /** Applies one already-ordered quote-critical event. */
  apply(cursor: ChainCursor, event: QuoteEvent): void {
    const previous = this.cursorValue;
    if (previous) {
      this.validateChainAndHash(previous, cursor);
      const previousIsHead =
        previous.blockNumber === cursor.blockNumber &&
        previous.transactionIndex === undefined &&
        previous.logIndex === undefined &&
        cursor.transactionIndex !== undefined &&
        cursor.logIndex !== undefined;
      const order = compareCursor(cursor, previous);
      if (!previousIsHead && order < 0) throw new ReducerError("CURSOR_REGRESSION", "cursor regression");
      if (!previousIsHead && order === 0) return;
    }
    this.validateEvent(event);
    this.applyEvent(event);
    this.cursorValue = { ...cursor };
  }

  /** Computes one quote synchronously from the current event-loop snapshot. */
  quote(request: QuoteRequest): QuoteOutcome {
    const cursor = this.requireReadyCursor();
    return computeQuote(request, cursor.executionBlockNumber, this.state);
  }

  /** Computes a batch without yielding, so every result uses one cursor/state. */
  quoteMany(requests: readonly QuoteRequest[]): readonly QuoteOutcome[] {
    const cursor = this.requireReadyCursor();
    return requests.map((request) => computeQuote(request, cursor.executionBlockNumber, this.state));
  }

  /** Creates a deep-cloned v3 restart checkpoint outside the quote path. */
  checkpoint(config: DeploymentConfig): Checkpoint | undefined {
    if (!this.cursorValue) return undefined;
    return {
      schemaVersion: SCHEMA_VERSION,
      mathCompatibilityVersion: MATH_COMPATIBILITY_VERSION,
      expectedRuntimeCodeHash: config.expectedRuntimeCodeHash,
      chainId: config.chainId,
      core: config.core,
      router: config.router,
      cursor: { ...this.cursorValue },
      state: cloneState(this.state),
    };
  }

  private requireReadyCursor(): ChainCursor {
    if (!this.ready) throw new IndexerError("NOT_READY", "indexer is not ready");
    if (!this.cursorValue) throw new IndexerError("NO_CURSOR", "indexer has no cursor");
    return this.cursorValue;
  }

  private validateChainAndHash(previous: ChainCursor, next: ChainCursor): void {
    if (previous.chainId !== next.chainId) throw new ReducerError("CHAIN_ID_MISMATCH", "cursor chain id mismatch");
    if (
      previous.blockNumber === next.blockNumber &&
      previous.blockHash &&
      next.blockHash &&
      previous.blockHash.toLowerCase() !== next.blockHash.toLowerCase() &&
      !isRealtimeProgression(previous, next)
    )
      throw new ReducerError("BLOCK_HASH_MISMATCH", "block hash mismatch");
  }

  private validateEvent(event: QuoteEvent): void {
    if (event.kind === "SlippageKSet" && (event.newK < 0n || event.newK > BPS))
      throw new ReducerError("INVALID_SLIPPAGE_K", "slippage K exceeds BPS");
    if ((event.kind === "PartnerInfoSet" || event.kind === "PartnerFeeSet") && (event.fee < 0n || event.fee > BPS))
      throw new ReducerError("INVALID_WIDTH", "partner fee exceeds BPS");
    if (
      (event.kind === "DepositExecuted" || event.kind === "WithdrawalExecuted") &&
      (event.principal < 0n || event.principal > U128_MAX)
    )
      throw new ReducerError("INVALID_WIDTH", "principal does not fit uint128");
  }

  private applyEvent(event: QuoteEvent): void {
    switch (event.kind) {
      case "LaneAdded": {
        const lane = this.lane(event.asset);
        lane.flags |= LaneFlags.Exists;
        this.setLane(event.asset, lane);
        break;
      }
      case "LaneRemoved":
        (this.state.lanes as Map<Address, LaneState>).delete(key(event.asset));
        break;
      case "LaneUpdated": {
        const lane = this.lane(event.asset);
        lane.slot0 = event.slot0;
        this.setLane(event.asset, lane);
        break;
      }
      case "SlippageKSet": {
        const lane = this.lane(event.asset);
        lane.slippageKBps = Number(event.newK);
        this.setLane(event.asset, lane);
        break;
      }
      case "PartnerInfoSet":
      case "PartnerFeeSet":
        if (key(event.router) === key(this.configuredRouter))
          (this.state.feeProfile.partnerFeeBps as Map<Address, number>).set(key(event.asset), Number(event.fee));
        break;
      case "WhitelistSet":
        if (key(event.router) === key(this.configuredRouter)) this.state.feeProfile.whitelisted = event.whitelisted;
        break;
      case "BlacklistFeeMultiplierSet":
        this.state.feeProfile.blacklistFeeMultiplier = event.multiplier;
        break;
      case "DepositExecuted": {
        const lane = this.lane(event.asset);
        const next = lane.totalPrincipalAmount + event.principal;
        if (next > U128_MAX) throw new ReducerError("ARITHMETIC", "principal storage overflow");
        lane.totalPrincipalAmount = next;
        this.setLane(event.asset, lane);
        break;
      }
      case "WithdrawalExecuted": {
        const lane = this.lane(event.asset);
        if (event.principal > lane.totalPrincipalAmount) throw new ReducerError("ARITHMETIC", "principal underflow");
        lane.totalPrincipalAmount -= event.principal;
        this.setLane(event.asset, lane);
        break;
      }
    }
  }

  private lane(asset: Address): LaneState {
    return { ...(this.state.lanes.get(key(asset)) ?? emptyLane()) };
  }

  private setLane(asset: Address, lane: LaneState): void {
    (this.state.lanes as Map<Address, LaneState>).set(key(asset), lane);
  }
}

function isRealtimeProgression(previous: ChainCursor, next: ChainCursor): boolean {
  return (
    previous.commitment === Commitment.Realtime &&
    next.commitment === Commitment.Realtime &&
    (next.sourceSequence ?? 0n) > (previous.sourceSequence ?? 0n)
  );
}

File: crates/lunarbase-client-core/src/state/reducer.rs
//! Ordered single-writer quote-state reducer.

use crate::model::{
    ChainCursor, Checkpoint, DeploymentConfig, MATH_COMPATIBILITY_VERSION, QuoteEvent,
    SCHEMA_VERSION,
};
use lunarbase_math::arithmetic::BPS;
use lunarbase_math::state::QuoteState;
use lunarbase_math::types::Address;
use thiserror::Error;

#[derive(Clone, Debug, Error, Eq, PartialEq)]
/// Failure that revokes quote readiness until canonical recovery.
pub enum ReducerError {
    /// An update cursor belongs to another EIP-155 chain.
    #[error("cursor chain id mismatch")]
    ChainIdMismatch,
    /// An event precedes the last accepted position in deterministic ordering.
    #[error("cursor regression")]
    CursorRegression,
    /// A provider retracted an applied log, requiring canonical state rebuild.
    #[error("removed log requires canonical rebuild")]
    RemovedLog,
    /// Two non-progressive cursors claim different hashes for one block height.
    #[error("block hash mismatch")]
    BlockHashMismatch,
    /// A state delta would underflow, overflow, or otherwise violate arithmetic invariants.
    #[error("arithmetic transition failed")]
    Arithmetic,
    /// A lane slippage coefficient exceeds the contract's basis-point bound.
    #[error("invalid lane slippage K")]
    InvalidSlippageK,
    /// An event value cannot fit the corresponding compact contract field.
    #[error("event value does not fit the contract storage width")]
    InvalidWidth,
}

#[derive(Clone, Debug)]
/// Single-writer reducer over quote-critical state.
///
/// Every fallible value is validated before mutation, avoiding the previous
/// full-state clone used for rollback on each event.
pub struct QuoteReducer {
    /// Complete quote-critical state mutated only by the ordered reducer task.
    state: QuoteState,
    /// Only router whose partner fee and whitelist changes affect this instance.
    configured_router: Address,
    /// Last normalized head or event position accepted by the reducer.
    cursor: Option<ChainCursor>,
    /// Fail-closed publication flag cleared on gaps, reorgs, and reducer errors.
    ready: bool,
}

impl QuoteReducer {
    /// Creates a not-ready reducer around a block-tagged state.
    pub fn new(state: QuoteState, configured_router: Address) -> Self {
        Self {
            state,
            configured_router,
            cursor: None,
            ready: false,
        }
    }

    /// Restores a ready reducer from a prevalidated checkpoint.
    pub fn from_checkpoint(checkpoint: Checkpoint) -> Self {
        Self {
            state: checkpoint.state,
            configured_router: checkpoint.router,
            cursor: Some(checkpoint.cursor),
            ready: true,
        }
    }

    /// Returns the current immutable state view.
    pub fn state(&self) -> &QuoteState {
        &self.state
    }

    /// Returns the last accepted cursor.
    pub fn cursor(&self) -> Option<&ChainCursor> {
        self.cursor.as_ref()
    }

    /// Reports whether the state may be used for quotes.
    pub const fn is_ready(&self) -> bool {
        self.ready
    }

    /// Revokes readiness until a canonical recovery succeeds.
    pub fn mark_not_ready(&mut self) {
        self.ready = false;
    }

    /// Publishes the current canonical state.
    pub fn publish_ready(&mut self) {
        self.ready = true;
    }

    /// Installs the initial snapshot cursor.
    pub fn bootstrap(&mut self, cursor: ChainCursor) {
        self.cursor = Some(cursor);
        self.ready = true;
    }

    /// Advances a block-level cursor without changing quote state.
    pub fn observe_head(&mut self, head: ChainCursor) -> Result<(), ReducerError> {
        if let Some(current) = &mut self.cursor {
            if current.chain_id != head.chain_id {
                return Err(ReducerError::ChainIdMismatch);
            }
            if current.block_number == head.block_number
                && current.block_hash.is_some()
                && head.block_hash.is_some()
                && current.block_hash != head.block_hash
                && !is_realtime_progression(current, &head)
            {
                return Err(ReducerError::BlockHashMismatch);
            }
            if head.block_number < current.block_number {
                return Ok(());
            }
            if head.block_number == current.block_number {
                if current.block_hash.is_none() || is_realtime_progression(current, &head) {
                    current.block_hash = head.block_hash;
                }
                current.execution_block_number = head.execution_block_number;
                if head.commitment > current.commitment {
                    current.commitment = head.commitment;
                }
                if head.source_sequence > current.source_sequence {
                    current.source_sequence = head.source_sequence;
                    current.source_sub_index = head.source_sub_index;
                }
                return Ok(());
            }
        }
        self.cursor = Some(head);
        Ok(())
    }

    /// Applies one decoded event after validating ordering and storage widths.
    pub fn apply(&mut self, cursor: ChainCursor, event: QuoteEvent) -> Result<(), ReducerError> {
        if let Some(previous) = &self.cursor {
            if previous.chain_id != cursor.chain_id {
                return Err(ReducerError::ChainIdMismatch);
            }
            if previous.block_number == cursor.block_number
                && previous.block_hash.is_some()
                && cursor.block_hash.is_some()
                && previous.block_hash != cursor.block_hash
                && !is_realtime_progression(previous, &cursor)
            {
                return Err(ReducerError::BlockHashMismatch);
            }
            let previous_is_block_head = previous.block_number == cursor.block_number
                && previous.transaction_index.is_none()
                && previous.log_index.is_none()
                && cursor.transaction_index.is_some()
                && cursor.log_index.is_some();
            if !previous_is_block_head && cursor.event_order() < previous.event_order() {
                return Err(ReducerError::CursorRegression);
            }
            if !previous_is_block_head && cursor.event_order() == previous.event_order() {
                return Ok(());
            }
        }
        self.apply_event(event)?;
        self.cursor = Some(cursor);
        Ok(())
    }

    fn apply_event(&mut self, event: QuoteEvent) -> Result<(), ReducerError> {
        match event {
            QuoteEvent::LaneAdded { asset } => {
                self.state.lanes.entry(asset).or_default().set_exists(true);
            }
            QuoteEvent::LaneRemoved { asset } => {
                self.state.lanes.remove(&asset);
            }
            QuoteEvent::LaneUpdated { asset, slot0 } => {
                self.state.lanes.entry(asset).or_default().slot0 = slot0;
            }
            QuoteEvent::SlippageKSet { asset, new_k } => {
                if new_k > BPS {
                    return Err(ReducerError::InvalidSlippageK);
                }
                let value = u32::try_from(new_k).map_err(|_| ReducerError::InvalidWidth)?;
                self.state.lanes.entry(asset).or_default().slippage_k_bps = value;
            }
            QuoteEvent::PartnerInfoSet { router, asset, fee }
            | QuoteEvent::PartnerFeeSet { router, asset, fee } => {
                if router != self.configured_router {
                    return Ok(());
                }
                if fee > BPS {
                    return Err(ReducerError::InvalidWidth);
                }
                let value = u32::try_from(fee).map_err(|_| ReducerError::InvalidWidth)?;
                self.state.fee_profile.partner_fee_bps.insert(asset, value);
            }
            QuoteEvent::WhitelistSet {
                router,
                whitelisted,
            } => {
                if router == self.configured_router {
                    self.state.fee_profile.whitelisted = whitelisted;
                }
            }
            QuoteEvent::BlacklistFeeMultiplierSet { multiplier } => {
                self.state.fee_profile.blacklist_fee_multiplier = multiplier;
            }
            QuoteEvent::DepositExecuted { asset, principal } => {
                let principal =
                    u128::try_from(principal).map_err(|_| ReducerError::InvalidWidth)?;
                let lane = self.state.lanes.entry(asset).or_default();
                let next = lane
                    .total_principal_amount
                    .checked_add(principal)
                    .ok_or(ReducerError::Arithmetic)?;
                lane.total_principal_amount = next;
            }
            QuoteEvent::WithdrawalExecuted { asset, principal } => {
                let principal =
                    u128::try_from(principal).map_err(|_| ReducerError::InvalidWidth)?;
                let lane = self.state.lanes.entry(asset).or_default();
                let next = lane
                    .total_principal_amount
                    .checked_sub(principal)
                    .ok_or(ReducerError::Arithmetic)?;
                lane.total_principal_amount = next;
            }
        }
        Ok(())
    }

    /// Builds a durable v3 checkpoint. This clone is outside the quote path.
    pub fn checkpoint(&self, deployment: &DeploymentConfig) -> Option<Checkpoint> {
        Some(Checkpoint {
            schema_version: SCHEMA_VERSION,
            math_compatibility_version: MATH_COMPATIBILITY_VERSION.into(),
            expected_runtime_code_hash: deployment.expected_runtime_code_hash,
            chain_id: deployment.chain_id,
            core: deployment.core,
            router: deployment.router,
            cursor: self.cursor.clone()?,
            state: self.state.clone(),
        })
    }
}

fn is_realtime_progression(previous: &ChainCursor, next: &ChainCursor) -> bool {
    previous.commitment == crate::model::Commitment::Realtime
        && next.commitment == crate::model::Commitment::Realtime
        && next.source_sequence > previous.source_sequence
}
